lambdas/pkgselect/index.py

```diff
@@ -90,172 +86,227 @@ def file_list_to_folder(df: pd.DataFrame, limit: int, offset: int) -> dict:
         objects = []
         total_results = 0
 
-    returned_results = len(prefixes) + len(objects)
     return dict(
         total=total_results,
-        returned=returned_results,
         prefixes=prefixes,
-        objects=objects
+        objects=objects,
     )
 
 
-def create_s3_client(
-    *,
-    aws_access_key_id: str,
-    aws_secret_access_key: str,
-    aws_session_token: str
-):
+def object_is_publicly_readable(bucket: str, key: str):
+    try:
+        boto3.client(
+            "s3",
+            config=botocore.client.Config(signature_version=botocore.UNSIGNED),
+        ).head_object(Bucket=bucket, Key=key)
+        return True
+    except botocore.exceptions.ClientError as ex:
+        if ex.response.get("Error", {}).get("Code") == "403":
+            return False
+        raise ex
+
+
+def create_s3_client(credentials: T.Optional[dict], bucket: str, key: str):
     """
-    Create an S3 Client using caller-provided credentials.
+    Create an S3 client for reading a manifest at a given bucket/key.
+    Use default session (backed by lambda role with GetObject permissions for
+    all buckets (*)) if the manifest is publicly accessible, otherwise
+    out-of-stack public buckets won't be accessible by logged-in users (since
+    they have S3 permissions only allowing to read specific buckets).
+    Use caller-provided credentials to start a session if the manifest is not
+    publicly accessible.
     """
-    assert aws_access_key_id and aws_secret_access_key and aws_session_token
-    session = boto3.Session(
-        aws_access_key_id=aws_access_key_id,
-        aws_secret_access_key=aws_secret_access_key,
-        aws_session_token=aws_session_token
+    if credentials is None and not os.getenv("ALLOW_ANONYMOUS_ACCESS"):
+        raise MissingCredentials
+
+    if (credentials is not None and not all((
+        credentials.get("aws_access_key_id"),
+        credentials.get("aws_secret_access_key"),
+        credentials.get("aws_session_token"),
+    ))):
+        raise IncompleteCredentials
+
+    if object_is_publicly_readable(bucket, key):
+        return boto3.client("s3")
+
+    if credentials is None:
+        raise AccessDenied
+
+    return boto3.Session(**credentials).client("s3")
```

Comment:
I think this also works:
```suggestion
    return boto3.client("s3", **credentials)
```